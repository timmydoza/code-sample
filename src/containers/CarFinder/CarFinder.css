.grid {
  max-width: 700px;
  margin: 0 auto;
}

@media (max-width: 750px) {
  .grid {
    width: 90%;
  }
}