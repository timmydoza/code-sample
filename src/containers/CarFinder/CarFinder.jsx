import React, { Component } from 'react';
import CarList from '../../components/CarList/CarList';
import CarModal from '../../components/CarModal/CarModal';
import styles from './CarFinder.css';
import API from '../../API/API';
import debounce from 'lodash.debounce';
import { getSortFn, getFilterFn } from '../../utils/utils';

const ITEMS_PER_PAGE = 15;

class CarFinder extends Component {

  state = {
    filteredSortedCars: [],
    sortOption: 'year',
    searchText: '',
    selectedCarKey: null,
    currentPage: 1,
    totalPages: null
  };

  carMasterList = [];

  selectCar = (selectedCarKey) => {
    this.setState({
      selectedCarKey
    });
  }

  setSort = (sortOption) => {
    this.setState(prevState => {
      return {
        sortOption,
        currentPage: 1,
        filteredSortedCars: this.sortAndFilter(this.carMasterList, sortOption, prevState.searchText)
      }
    });
  }

  setSearch = debounce(searchText => {
    this.setState(prevState => {
      const filteredSortedCars = this.sortAndFilter(this.carMasterList, prevState.sortOption, searchText);
      return {
        searchText,
        currentPage: 1,
        filteredSortedCars,
        totalPages: this.getTotalPages(filteredSortedCars)
      }
    });
  }, 50)

  componentDidMount = () => {
    API().then(cars => {
      this.carMasterList = cars;
      this.setState(prevState => {
        const filteredSortedCars = this.sortAndFilter(cars, prevState.sortOption, prevState.searchText);
        return {
          filteredSortedCars,
          totalPages: this.getTotalPages(filteredSortedCars)
        }
      });
    });
  }

  setPage = (pageChange) => {
    this.setState(prevState => {
      let newPage = prevState.currentPage + pageChange;

      if (newPage < 1) {
        newPage = 1;
      }

      if (newPage > prevState.totalPages) {
        newPage = prevState.totalPages;
      }

      return {
        currentPage: newPage
      }
    });
  }

  sortAndFilter(carArray, sortOption, searchText) {
    //Filter returns new array.  Being careful to not mutate this.state.
    const cars = carArray.filter(getFilterFn(searchText));
    return cars.sort(getSortFn(sortOption));
  }

  getTotalPages = carList => {
    return Math.ceil(carList.length / ITEMS_PER_PAGE);
  }

  paginate = carList => {
    const page = this.state.currentPage;
    return carList.slice((page - 1) * ITEMS_PER_PAGE, page * ITEMS_PER_PAGE);
  }

  getSelectedCar = () => {
    return this.state.filteredSortedCars.find(car => car.key === this.state.selectedCarKey);
  }

  render() {

    return (
      <main className={styles.grid}>
        <CarList 
          cars={this.paginate(this.state.filteredSortedCars)}
          selectCar={this.selectCar}
          selectedCarKey={this.state.selectedCarKey}
          setSort={this.setSort}
          setSearch={this.setSearch}
          currentPage={this.state.currentPage}
          totalPages={this.state.totalPages}
          setPage={this.setPage}
        />
        <CarModal 
          { ...this.getSelectedCar() } 
          selectCar={this.selectCar} 
          selectedCarKey={this.state.selectedCarKey} />
      </main>
    );
  }
}

export default CarFinder;
