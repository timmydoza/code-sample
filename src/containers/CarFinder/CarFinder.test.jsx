import React from 'react';
import renderer from 'react-test-renderer';
import CarFinder from './CarFinder';

jest.mock('../../API/API');

describe('the CarFinder component', () => {
  let carFinder, setPage;

  beforeEach(() => {
    carFinder = renderer.create(<CarFinder />);
    debugger;
  });

  it('renders correctly', () => {
    expect(carFinder.toJSON()).toMatchSnapshot();
  });

});