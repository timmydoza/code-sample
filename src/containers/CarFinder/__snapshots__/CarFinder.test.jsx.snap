// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the CarFinder component renders correctly 1`] = `
<main
  className="grid"
>
  <div>
    <div
      className="filterContainer"
    >
      <div
        className="sortControl"
      >
        <label>
          Sort By:
          <select
            onChange={[Function]}
          >
            <option
              value="year"
            >
              Year - Oldest to Newest
            </option>
            <option
              value="mileage"
            >
              Mileage - Highest to Lowest
            </option>
            <option
              value="created_at"
            >
              Date Listed - Newest to Oldest
            </option>
          </select>
        </label>
      </div>
      <div>
        <span>
          Search: 
        </span>
        <input
          onChange={[Function]}
          type="text"
        />
      </div>
    </div>
    <table
      className="CarList"
    >
      <thead>
        <tr>
          <th>
            Year
          </th>
          <th>
            Make
          </th>
          <th>
            Model
          </th>
          <th>
            Mileage
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            1999
          </td>
          <td>
            BMW
          </td>
          <td>
            528i
          </td>
          <td>
            160254
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2013
          </td>
          <td>
            Jeep
          </td>
          <td>
            Wrangler Unlimited
          </td>
          <td>
            19000
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2013
          </td>
          <td>
            Hyundai
          </td>
          <td>
            Accent
          </td>
          <td>
            21587
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2013
          </td>
          <td>
            Kia
          </td>
          <td>
            Optima
          </td>
          <td>
            24235
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2013
          </td>
          <td>
            Nissan
          </td>
          <td>
            Maxima
          </td>
          <td>
            17489
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2013
          </td>
          <td>
            Volkswagen
          </td>
          <td>
            Eos
          </td>
          <td>
            18145
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2013
          </td>
          <td>
            Audi
          </td>
          <td>
            A5
          </td>
          <td>
            17216
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2014
          </td>
          <td>
            Nissan
          </td>
          <td>
            Titan
          </td>
          <td>
            15478
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2014
          </td>
          <td>
            Honda
          </td>
          <td>
            Civic
          </td>
          <td>
            21458
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2014
          </td>
          <td>
            Jaguar
          </td>
          <td>
            XK
          </td>
          <td>
            9852
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2014
          </td>
          <td>
            Nissan
          </td>
          <td>
            Juke
          </td>
          <td>
            10457
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2014
          </td>
          <td>
            SRT
          </td>
          <td>
            Viper
          </td>
          <td>
            8411
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2014
          </td>
          <td>
            Mini
          </td>
          <td>
            Cooper
          </td>
          <td>
            12457
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2014
          </td>
          <td>
            Infinity
          </td>
          <td>
            Q60
          </td>
          <td>
            9256
          </td>
        </tr>
        <tr
          className="CarRow"
          onClick={[Function]}
        >
          <td>
            2014
          </td>
          <td>
            Buick
          </td>
          <td>
            Enclave
          </td>
          <td>
            9587
          </td>
        </tr>
      </tbody>
    </table>
    <div
      className="paginationButtons"
    >
      <button
        onClick={[Function]}
      >
        &lt;
      </button>
      <span>
        1
         of 
        2
      </span>
      <button
        onClick={[Function]}
      >
        &gt;
      </button>
    </div>
  </div>
  <div
    className="modalOverlay"
    onClick={[Function]}
  >
    <div
      className="modal"
      onClick={[Function]}
    >
      <span
        className="exit"
        onClick={[Function]}
      >
        X
      </span>
      <ul>
        <li>
          Year: 
        </li>
        <li>
          Make: 
        </li>
        <li>
          Model: 
        </li>
        <li>
          Mileage: 
        </li>
      </ul>
      <img
        alt=""
        src={undefined}
      />
    </div>
  </div>
</main>
`;
